src/v3/src/transformer/selectAuthenticator/transformSelectAuthenticatorEnroll.ts

```diff
@@ -64,39 +94,107 @@ export const transformSelectAuthenticatorEnroll: IdxStepTransformer = ({
       content: loc('oie.select.authenticators.enroll.title', 'login'),
     },
   };
-  const informationalText: DescriptionElement = {
+  const description: DescriptionElement = {
     type: 'Description',
     contentType: 'subtitle',
     options: getContentDescrAndParams(brandName),
   };
-  const description: HeadingElement = {
+
+  const headingRequiredNow: HeadingElement = {
+    type: 'Heading',
+    options: {
+      content: loc('oie.setup.required.now', 'login'),
+      level: 3,
+      visualLevel: 6,
+      dataSe: 'authenticator-list-title',
+    },
+  };
+
+  const headingRequiredSoon: HeadingElement = {
+    type: 'Heading',
+    options: {
+      content: loc('oie.setup.required.soon', 'login'),
+      level: 3,
+      visualLevel: 6,
+      dataSe: 'authenticator-list-title',
+    },
+  };
+
+  const headingNoGracePeriod: HeadingElement = {
     type: 'Heading',
     options: {
-      content: skipStep ? loc('oie.setup.optional', 'login') : loc('oie.setup.required', 'login'),
+      content: skipStep ? loc('oie.setup.optional.short', 'login') : loc('oie.setup.required.now', 'login'),
       level: 3,
       visualLevel: 6,
     },
   };
-  const authenticatorListElement: AuthenticatorButtonListElement = {
-    type: 'AuthenticatorButtonList',
-    options: { buttons: authenticatorButtons, dataSe: 'authenticator-enroll-list' },
+
+  const descriptionGracePeriod: DescriptionElement = {
+    type: 'Description',
+    contentType: 'subtitle',
+    options: {
+      content: loc('oie.setup.required.soon.description', 'login'),
+    },
   };
+
+  const authenticatorListElementWithGracePeriod: AuthenticatorButtonListElement[] = [];
+  if (authenticatorButtonsWithGracePeriod.length) {
+    authenticatorListElementWithGracePeriod.push({
+      type: 'AuthenticatorButtonList',
+      options: { buttons: authenticatorButtonsWithGracePeriod, dataSe: 'authenticator-enroll-list-grace-period' },
+    });
+  }
+
+  const authenticatorListElementDueNow: AuthenticatorButtonListElement[] = [];
+  if (authenticatorButtonsDueNow.length) {
+    authenticatorListElementDueNow.push({
+      type: 'AuthenticatorButtonList',
+      options: { buttons: authenticatorButtonsDueNow, dataSe: 'authenticator-enroll-list' },
+    });
+  }
+
   const skipButton: ButtonElement = {
     type: 'Button',
-    label: loc('oie.optional.authenticator.button.title', 'login'),
+    label: authenticatorListElementWithGracePeriod.length
+      ? loc('oie.setup.remind.me.later', 'login')
+      : loc('oie.optional.authenticator.button.title', 'login'),
     options: {
       type: ButtonType.SUBMIT,
       step: 'skip',
     },
   };
 
-  uischema.elements = [
-    title,
-    informationalText,
-    description,
-    authenticatorListElement,
-    ...(skipStep ? [skipButton] : []),
-  ];
+  // 3 situations - required soon + required now, all required soon, all required now
+  // when grace periods are past required and should be treated as normal required
```

Comment:
Could this be handled by sequentially building the array and assigning it to `uischema.elements` at the end? Like this:
```js
const elements = [title, description];
if (hasNow || hasSoon) {
  if (hasNow) {
    // add now heading, elements
  }
  if (hasSoon) {
    // add soon heading, elements
  }
} else {
  // add default heading, elements
}
if (skipStep) {
  // add skipButton
}
ui.elements = elements;
```

Reply:
Yup, updated!